frame/society/src/migrations.rs

```diff
@@ -50,16 +53,16 @@ impl<
 	}
 
 	fn on_runtime_upgrade() -> Weight {
-		let current = Pallet::<T, I>::current_storage_version();
 		let onchain = Pallet::<T, I>::on_chain_storage_version();
-		if current == 2 && onchain == 0 {
-			from_original::<T, I>(&mut PastPayouts::get())
-		} else {
+		if onchain < 2 {
 			log::info!(
-				"Running migration with current storage version {:?} / onchain {:?}",
-				current,
+				target: TARGET,
+				"Running migration against onchain version {:?}",
 				onchain
 			);
+			from_original::<T, I>(&mut PastPayouts::get()).defensive_unwrap_or(Weight::MAX)
+		} else {
+			defensive!("Unexpected onchain version: {:?}. Expected 0.");
```

Comment:
But this should not happen anymore since we wrap it in a `VersionedRuntimeUpgrade`.

Reply:
Only when you enable the experimental feature. But still, just because there is maybe some way to prevent this, I don't see any reason we can not fix it at the root of the problem.

Reply:
I'll change it back to a log::warn.

Reply:
Keep in mind, `VersionedRuntimeUpgrade` is intended to allow developers to not need to worry about any version checking in their `OnRuntimeUpgrade` hooks. 

So when it is out of experimental, we need some way to mark an upgrade in a way so it can only be used when wrapped in `VersionedRuntimeUpgrade`.

Reply:
> So when it is out of experimental, we need some way to mark an upgrade in a way so it can only be used when wrapped in `VersionedRuntimeUpgrade`.

You signal this by not making the "inner migration" public and only export the wrapped `VersionedRuntimeUpgrade` type.

Reply:
I already tried removing pub from the `Unchecked` migration actually but it results in a compiler error:

```
error[E0446]: private type `VersionUncheckedMigrateToV2<T, I, PastPayouts>` in public interface
   --> frame/society/src/migrations.rs:102:1
    |
31  | struct VersionUncheckedMigrateToV2<T: Config<I>, I: 'static, PastPayouts>(
    | ------------------------------------------------------------------------- `VersionUncheckedMigrateToV2<T, I, PastPayouts>` declared as private
...
102 | pub type VersionCheckedMigrateToV2<T, I, PastPayouts> =
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak private type
```

I tried hacking on this a little and the only way I could make it work was via a wrapper struct, but it's really verbose and ugly:

```rust
type VersionCheckedMigrateToV2Type<T, I, PastPayouts> =
	frame_support::migrations::VersionedRuntimeUpgrade<
		0,
		2,
		VersionUncheckedMigrateToV2<T, I, PastPayouts>,
		crate::pallet::Pallet<T, I>,
		<T as frame_system::Config>::DbWeight,
	>;

pub struct VersionCheckedMigrateToV2<
	T: Config<I>,
	I: Instance + 'static,
	PastPayouts: Get<Vec<(<T as frame_system::Config>::AccountId, BalanceOf<T, I>)>>,
>(core::marker::PhantomData<(T, I, PastPayouts)>);

impl<
		T: Config<I>,
		I: Instance + 'static,
		PastPayouts: Get<Vec<(<T as frame_system::Config>::AccountId, BalanceOf<T, I>)>>,
	> OnRuntimeUpgrade for VersionCheckedMigrateToV2<T, I, PastPayouts>
{
	fn on_runtime_upgrade() -> Weight {
		VersionCheckedMigrateToV2Type::<T, I, PastPayouts>::on_runtime_upgrade()
	}

	#[cfg(feature = "try-runtime")]
	fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
		VersionCheckedMigrateToV2Type::<T, I, PastPayouts>::pre_upgrade()
	}

	#[cfg(feature = "try-runtime")]
	fn post_upgrade(_state: Vec<u8>) -> Result<(), TryRuntimeError> {
		VersionCheckedMigrateToV2Type::<T, I, PastPayouts>::post_upgrade(_state)
	}

	#[cfg(feature = "try-runtime")]
	fn try_on_runtime_upgrade(checks: bool) -> Result<Weight, TryRuntimeError> {
		VersionCheckedMigrateToV2Type::<T, I, PastPayouts>::try_on_runtime_upgrade(checks)
	}
}
```

Could be a good candidate for a macro, but I would like to avoid that and keep it vanilla Rust for the sake of simplicity if we can.

Let me know if you have any other ideas & if I may be missing something easy, otherwise I'll create an issue to track establishing a pattern and updating examples.

Reply:
Yeah the type needs to be public when it appears in the generic. However, you can make the type public but not export it from the `migrations` module ;)